File: May-LeetCoding-Challange-2020/Week3/1277-count-square-submatrices-with-all-ones.py
class Solution:
    # Sol 1. DP
    def countSquares(self, matrix: List[List[int]]) -> int:
        n = len(matrix) # # of rows
        print(n)
        m = len(matrix[0]) # # of cols
        print(m)
        count = 0
        dp_matrix = [[0 for j in range(m+1)] for i in range(n+1)]
        print(dp_matrix)
        for row in range(1,n+1):
            for col in range(1, m+1):
                if matrix[row-1][col-1] == 1:
                    dp_matrix[row][col] = 1+ min(dp_matrix[row][col-1], dp_matrix[row-1][col], dp_matrix[row-1][col-1])
                    count += dp_matrix[row][col]
                    
        return count
        # Time Complexity: O(n^2)
        # Space Complexity: O(nm)
        # Run Code Runtime: 48 ms
        # Submit Runtime: 724 ms, 35.47% faster
        # Submti Memory: 15.8 mb, 100% less 

File: May-LeetCoding-Challange-2020/Week3/1277-count-square-submatrices-with-all-ones.cpp
// Time Complexity: O(m*n)
// Space Compleixty: O(1)
// Runtime: 148 ms, faster than 34.48% 
// Memory Usage: 23.7 MB, less than 100.00%
class Solution {
public:
    int countSquares(vector<vector<int>>& matrix) {
        
        if(matrix.size() == 0)
            return 0;
        
        int count = 0;

        for(int i = 1; i < matrix.size(); i++){
            for(int j = 1; j < matrix[0].size(); j++){
                
                if(matrix[i][j] == 1){
                    matrix[i][j] += min(matrix[i-1][j-1],min(matrix[i-1][j], matrix[i][j-1]));
                    count += matrix[i][j];
                }
            }
        }
        
        count += matrix[0][0];
        
        for(int i = 1; i < matrix.size(); i++)
            count += matrix[i][0];
        
        
        for(int i = 1; i < matrix[0].size(); i++)
            count += matrix[0][i];
        
        return count;
    }
};